lightning/checkpoints/checkpoints.go

```diff
@@ -1199,6 +1205,54 @@ func (cpdb *MySQLCheckpointsDB) MoveCheckpoints(ctx context.Context, taskID int6
 	return nil
 }
 
+func (cpdb *MySQLCheckpointsDB) GetLocalStoringTables(ctx context.Context) (map[string][]int32, error) {
+	var targetTables map[string][]int32
+
+	// lightning didn't check CheckpointStatusMaxInvalid before this function is called, so we skip invalid ones
+	// engines should exist if
+	// 1. table status is earlier than CheckpointStatusIndexImported, and
+	// 2. engine status is earlier than CheckpointStatusImported, and
+	// 3. chunk has been read
+	query := fmt.Sprintf(`
+		SELECT t.table_name, c.engine_id 
+		FROM %s.%s t, %s.%s c, %s.%s e 
+		WHERE t.table_name = c.table_name AND t.table_name = e.table_name AND c.engine_id = e.engine_id 
+			AND %d < t.status AND t.status < %d 
+			AND %d < e.status AND e.status < %d
+			AND c.pos > c.offset;`,
+		cpdb.schema, CheckpointTableNameTable, cpdb.schema, CheckpointTableNameChunk, cpdb.schema, CheckpointTableNameEngine,
+		CheckpointStatusMaxInvalid, CheckpointStatusIndexImported,
+		CheckpointStatusMaxInvalid, CheckpointStatusImported)
+
+	err := common.Retry("get local storing tables", log.L(), func() error {
+		targetTables = make(map[string][]int32)
+		rows, err := cpdb.db.QueryContext(ctx, query)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var (
+				tableName string
+				engineID  int32
+			)
+			if err := rows.Scan(&tableName, &engineID); err != nil {
+				return errors.Trace(err)
+			}
+			targetTables[tableName] = append(targetTables[tableName], engineID)
```

Comment:
If there are multi chunks that are not finished, the append will likely execute multi times for the same engine id

Reply:
fixed in https://github.com/pingcap/tidb-lightning/pull/491/commits/979062c24a8cdf3a498bd482045f4a171657d229